Assets/Mirage/Runtime/InterestManagement/InterestManager.cs

```diff
@@ -0,0 +1,286 @@
+
+            stopWatch.Stop();
+
+            if (logger.logEnabled)
+                logger.Log($"[Interest Manager] - Send Method Execution Time: {stopWatch.Elapsed.TotalMilliseconds} ms");
+        }
+
+        /// <summary>
+        /// Send a message to all observers of an identity
+        /// </summary>
+        /// <remarks>Override if you wish to provide
+        /// an allocation free send method</remarks>
+        /// <param name="identity">the object that wants to send a message</param>
+        /// <param name="data">the data to send</param>
+        /// <param name="channelId">the channel to send it on</param>
+        /// <param name="skip">a player who should not receive the message</param>
+        /// <returns>Total amounts of messages sent</returns>
+        protected virtual int Send(NetworkIdentity identity, ArraySegment<byte> data, int channelId = Channel.Reliable, INetworkPlayer skip = null)
+        {
+            int count = 0;
+
+            foreach (INetworkPlayer player in Observers(identity))
+            {
+                if (player == null) continue;
+
+                if (player != skip)
+                {
+                    // send to all connections, but don't wait for them
+                    player.Send(data, channelId);
+                    count++;
+                }
+            }
+
+            return count;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void CheckGrowVisibilitySystem()
+        {
+            var stopWatch = Stopwatch.StartNew();
+
+            if (_visibilitySystems == null)
+            {
+                _visibilitySystems = new ObserverData[_initialSystems];
+            }
+            else if (_visibilitySystems[_visibilitySystems.Length - 1].System == default)
+            {
+                // We need to copy and expand our visibility system.
+                if (logger.logEnabled)
+                    logger.Log("[Interest Manager] - Visibility system is expanding array. If this is happening often. Please set initial systems in constructor.");
+
+                var newData = new ObserverData[_initialSystems + _initialSystems];
+
+                _visibilitySystems.CopyTo(newData, 0);
+
+                _visibilitySystems = newData;
+            }
+
+            stopWatch.Stop();
+
+            if (logger.logEnabled)
+                logger.Log($"[Interest Manager] - CheckGrowVisibilitySystem Method Execution Time: {stopWatch.Elapsed.TotalMilliseconds} ms");
+        }
+
+        /// <summary>
+        ///     Register a specific interest management system to the interest manager.
+        /// </summary>
+        /// <param name="system">The system we want to register in the interest manager.</param>
+        internal void RegisterVisibilitySystem(ref ObserverData system)
+        {
+            CheckGrowVisibilitySystem();
+
+            var stopWatch = Stopwatch.StartNew();
+
+            if (_visibilitySystems.Equals(system))
```

Comment:
should this be contains?

Reply:
Its array of data. There is no contains. That s why I created IEqualityComparer methods

Reply:
I havn't checked in a code editor but this loops like it is checks if the array is equal to the new item